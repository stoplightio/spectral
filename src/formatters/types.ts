import { IRuleResult } from '../types';

export type Formatter = (results: IRuleResult[]) => string;
