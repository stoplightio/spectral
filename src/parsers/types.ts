import { GetLocationForJsonPath, IParserResult } from '@stoplight/types';

export interface IParser<R extends IParserResult = IParserResult<unknown>> {
  parse(input: string): R;
  getLocationForJsonPath: GetLocationForJsonPath<R>;
  trapAccess<T extends object = object>(obj: T): T;
}
