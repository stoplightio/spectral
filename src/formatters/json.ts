import { IRuleResult } from '../types';

export const json = (results: IRuleResult[]): string => {
  const outputJson = results.map(result => {
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
