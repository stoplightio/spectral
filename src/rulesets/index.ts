export { mergeRulesets } from './merger';
export { readRulesFromRulesets } from './reader';
export { assertValidRuleset } from './validation';
