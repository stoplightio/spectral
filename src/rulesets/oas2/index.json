{
  "extends": ["spectral:oas"],
  "formats": ["oas2"],
  "rules": {
    "api-host": {
      "description": "OpenAPI `host` must be present and non-empty string.",
      "recommended": true,
      "type": "style",
      "given": "$",
      "then": {
        "field": "host",
        "function": "truthy"
      },
      "tags": [
        "api"
      ]
    },
    "api-schemes": {
      "description": "OpenAPI host `schemes` must be present and non-empty array.",
      "recommended": true,
      "type": "style",
      "given": "$",
      "then": {
        "field": "schemes",
        "function": "schema",
        "functionOptions": {
          "schema": {
            "items": {
              "type": "string"
            },
            "minItems": 1,
            "type": "array"
          }
        }
      },
      "tags": [
        "api"
      ]
    },
    "host-not-example": {
      "description": "Server URL should not point at `example.com`.",
      "recommended": false,
      "type": "style",
      "given": "$",
      "then": {
        "field": "host",
        "function": "pattern",
        "functionOptions": {
          "notMatch": "example\\.com"
        }
      }
    },
    "host-trailing-slash": {
      "description": "Server URL should not have a trailing slash.",
      "recommended": true,
      "type": "style",
      "given": "$",
      "then": {
        "field": "host",
        "function": "pattern",
        "functionOptions": {
          "notMatch": "/$"
        }
      }
    },
    "model-description": {
      "description": "Definition `description` must be present and non-empty string.",
      "recommended": false,
      "type": "style",
      "given": "$..definitions[*]",
      "then": {
        "field": "description",
        "function": "truthy"
      }
    },
    "oas2-parameter-description": {
      "description": "Parameter objects should have a `description`.",
      "recommended": false,
      "type": "style",
      "given": "$..[?(@parentProperty === 'parameters' && @.in)]",
      "then": {
        "field": "description",
        "function": "truthy"
      },
      "tags": [
        "parameters"
      ]
    },
    "oas2-operation-security-defined": {
      "description": "Operation `security` values must match a scheme defined in the `securityDefinitions` object.",
      "recommended": true,
      "type": "validation",
      "given": "$",
      "then": {
        "function": "oasOpSecurityDefined",
        "functionOptions": {
          "schemesPath": [
            "securityDefinitions"
          ]
        }
      },
      "tags": [
        "operation"
      ]
    },
    "valid-example-in-parameters": {
      "description": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "recommended": true,
      "severity": 0,
      "type": "validation",
      "given": "$..parameters..[?(@.in == 'body')]..[?(@property !== 'properties' && @.example && ( @.type || @.format || @.$ref ))]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$"
        }
      }
    },
    "valid-example-in-definitions": {
      "description": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "recommended": true,
      "severity": 0,
      "type": "validation",
      "given": "$..definitions..[?(@property !== 'properties' && @.example && (@.type || @.format || @.$ref))]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$"
        }
      }
    },
    "oas2-anyOf": {
      "description": "OpenAPI v3 keyword `anyOf` detected in OpenAPI v2 document.",
      "message": "anyOf is not available in OpenAPI v2, it was added in OpenAPI v3",
      "recommended": true,
      "type": "validation",
      "given": "$..[?(@.anyOf)]",
      "then": {
        "field": "anyOf",
        "function": "undefined"
      },
      "tags": [
        "schema"
      ]
    },
    "oas2-oneOf": {
      "description": "OpenAPI v3 keyword `oneOf` detected in OpenAPI v2 document.",
      "message": "oneOf is not available in OpenAPI v2, it was added in OpenAPI v3",
      "recommended": true,
      "type": "validation",
      "given": "$..[?(@.oneOf)]",
      "then": {
        "field": "oneOf",
        "function": "undefined"
      },
      "tags": [
        "schema"
      ]
    },
    "oas2-schema": {
      "description": "Validate structure of OpenAPI v2 specification.",
      "message": "{{error}}",
      "type": "validation",
      "severity": 0,
      "recommended": true,
      "given": "$",
      "then": {
        "function": "schema",
        "functionOptions": {
          "schema": {
            "$ref": "./schemas/main.json"
          }
        }
      },
      "tags": [
        "schema"
      ]
    }
  }
}
