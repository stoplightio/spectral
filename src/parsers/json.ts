import { getLocationForJsonPath, JsonParserResult, parseWithPointers, trapAccess } from '@stoplight/json';
import { IParser } from './types';

export const parseJson = (input: string) =>
  parseWithPointers(input, { ignoreDuplicateKeys: false, preserveKeyOrder: true });

export const Json: IParser<JsonParserResult<unknown>> = {
  parse: parseJson,
  getLocationForJsonPath,
  trapAccess,
};
