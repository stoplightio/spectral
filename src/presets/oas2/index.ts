const merge = require('lodash.merge');

import { IPreset } from '../../types';
import { preset as commonOas } from '../oas';
import * as schema from './schemas/main.json';

export const oas2Preset = (): IPreset => {
  return {
    name: 'oas2',
    rules: merge(commonOas().rules, {
      oas2: {
        'oas2-schema': {
          type: 'validation',
          summary: 'Validate structure of OpenAPIv2 specification.',
          enabled: true,
          severity: 'error',
          path: '$',
          function: 'schema',
          input: {
            schema,
          },
        },
      },
    }),
  };
};
