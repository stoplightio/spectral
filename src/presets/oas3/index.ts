const merge = require('lodash.merge');

import { IPreset } from '../../types';
import { preset as commonOas } from '../oas';
import * as schema from './schemas/main.json';

export const oas3Preset = (): IPreset => {
  return {
    name: 'oas3',
    rules: merge(commonOas().rules, {
      oas3: {
        'oas3-schema': {
          type: 'validation',
          summary: 'Validate structure of OpenAPIv3 specification.',
          enabled: true,
          severity: 'error',
          path: '$',
          function: 'schema',
          input: {
            schema,
          },
        },
      },
    }),
  };
};
