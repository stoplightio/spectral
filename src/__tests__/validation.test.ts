import { oas2Functions, oas2Rules } from '../rulesets/oas2';
import { oas3Functions, oas3Rules } from '../rulesets/oas3';
import { Spectral } from '../spectral';
import { RuleType } from '../types';
import * as petstoreV2 from './fixtures/petstore.oas2.json';
import * as petstoreV3 from './fixtures/petstore.oas3.json';

const invalidV2 = require('./fixtures/todos.invalid.oas2.json');
const oas2Ruleset = {
  functions: oas2Functions(),
  rules: oas2Rules(),
};
const oas3Ruleset = {
  functions: oas3Functions(),
  rules: oas3Rules(),
};

describe('validation', () => {
  test('validate a correct OASv2 spec', () => {
    const s = new Spectral();
    s.setFunctions(oas2Ruleset.functions || {});
    s.setRules(oas2Ruleset.rules);
    const result = s.run(petstoreV2, { format: 'oas2', type: RuleType.VALIDATION });
    expect(result.results.length).toEqual(0);
  });

  test('return errors on invalid OASv2 spec', () => {
    const s = new Spectral();
    s.setFunctions(oas2Ruleset.functions || {});
    s.setRules(oas2Ruleset.rules);
    const result = s.run(invalidV2, { format: 'oas2', type: RuleType.VALIDATION });
    expect(result.results.length).toEqual(1);
    expect(result.results[0].path).toEqual(['$', 'info', 'license', 'name']);
    expect(result.results[0].message).toEqual('should be string');
  });

  test('validate a correct OASv3 spec', () => {
    const s = new Spectral();
    s.setFunctions(oas3Ruleset.functions || {});
    s.setRules(oas3Ruleset.rules);
    const result = s.run(petstoreV3, { format: 'oas3', type: RuleType.VALIDATION });
    expect(result.results.length).toEqual(0);
  });

  test('validate multiple formats with same validator', () => {
    const s = new Spectral();
    s.setFunctions(oas2Ruleset.functions || {});
    s.setRules(oas2Ruleset.rules);
    s.mergeFunctions(oas3Ruleset.functions || {});
    s.mergeRules(oas3Ruleset.rules);

    let result = s.run(petstoreV2, { format: 'oas2', type: RuleType.VALIDATION });
    expect(result.results.length).toEqual(0);

    result = s.run(invalidV2, { format: 'oas2', type: RuleType.VALIDATION });
    expect(result.results).toMatchSnapshot();

    result = s.run(petstoreV3, { format: 'oas3', type: RuleType.VALIDATION });
    expect(result.results.length).toEqual(0);
  });
});
