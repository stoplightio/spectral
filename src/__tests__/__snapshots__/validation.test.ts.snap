// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation validate multiple formats with same validator 1`] = `
Array [
  Object {
    "message": "should be string",
    "path": Array [
      "$",
      "info",
      "license",
      "name",
    ],
  },
]
`;
